### DISCUSSION TITLE
Setup For Subclassed Association Table With Polymorphic Entities

### DISCUSSION THREAD

**Discussion by coneybeare** [sqlalchemy/sqlalchemy#6862] [Usage Questions] [answered]
(cross-posted on [Stack Overflow](https://stackoverflow.com/questions/68671417/sqlalchemy-setup-for-association-table-with-polymorphic-entities))

I am new to both `python` and `sqlalchemy` but have a rather complex modeling situation I am having trouble setting up. It involves a subclassed association table in which the association table has a polymorphic association for one of the entities.

I am *very* close to getting this working. When the data already exists in the DB, I can read and interact with it and the models as expected. The problem comes with writing, and I will address that after showing the code:

First, there is a shared base class defining tablename and id as postgres uuid

```python
@as_declarative()
class Base(object):
    @declared_attr
    def __tablename__(cls):
        return cls.__name__.lower()

    id = Column(pg.UUID(as_uuid=True), primary_key=True, default=uuid4)

```

There is an enum of defined types allowed in the polymorphic column. While I will eventually support Aaa and Bbb, this example only has Bbb defined so far for clarity.

```
class EntityTypes(Enum):
    AAA = Aaa.__name__.lower()
    BBB = Bbb.__name__.lower()
```

This is the model that represents the polymorphic association table. It has `entity_id`, `entity_type`, and `ccc_id` columns for the joins. "Entity" can be an Aaa or a Bbb, but Ccc is always a Ccc (non-polymorphic).

```python
class EntityCcc(Base):
    """Polymorphic mapping between an EntityType and Ccc."""
    
    # entity_id is defined in subclasses, with foreign keys
    entity_type = Column(
        Enum(EntityTypes, values_callable=lambda x: [e.value for e in x]), nullable=False
    )

    ccc_id = Column(
        pg.UUID(as_uuid=True),
        ForeignKey(f"{Ccc.__tablename__}.id"),
        nullable=False
    )

    __mapper_args__ = {
        "polymorphic_on": entity_type
    }
```

This is a subclass of the polymorphic model which knows about `Bbb`s, so it sets the foreign key accordingly, and sets up the relationships. A similar `AaaCcc` class would be defined

```python
class BbbCcc(EntityCcc):
    """
        Mapping between a Bbb and Ccc.
        Subclasses polymorphic join model to get specific
        bbb accessor and set appropriate foreign key.
    """

    __tablename__ = EntityCcc.__tablename__

    entity_id = Column(
        pg.UUID(as_uuid=True), ForeignKey(f"{Bbb.__tablename__}.id"), nullable=False
    )

    bbb = relationship("Bbb", back_populates='bbb_ccc')
    ccc = relationship("Ccc", back_populates='bbb_ccc')

    __mapper_args__ = {
        "polymorphic_identity": EntityTypes(Bbb.__name__.lower())
    }
```

This is a `Bbb`, which has a relationship setup to the subclassed join model, and a secondary relationship setup to point to its `Ccc` through the join table (using `secondary`)

```python
class Bbb(Base):
    """Represents a Bbb entity."""

    name = Column(TEXT)

    bbb_ccc = relationship("BbbCcc", back_populates="bbb", uselist=False)

    ccc = relationship(
        "Ccc",
        secondary="entity_ccc",
        back_populates="bbb",
        uselist=False
    )

```

This is a Ccc, which has a relationship setup to the subclassed join model, and a secondary relationship setup to point to its `Bbb` through the join table (using `secondary`)

```python
class Ccc(Base):
    """Represents a Ccc entity."""

    name = Column(TEXT)

    bbb_ccc = relationship("BbbCcc", back_populates="ccc", uselist=False)

    bbb = relationship(
        "Bbb",
        secondary="entity_ccc",
        back_populates="ccc",
        uselist=False
    )
```

<hr>

**So what's the problem?**

With the appropriate entries already seeded in the database, I can interact with them as expected:

```python
(Pdb) found_bbb_ccc = db.session.query(BbbCcc).get(uuid)
(Pdb) found_bbb_ccc
<app.models.mappings.bbb_ccc.BbbCcc object at 0x7f488ce6ebe0>
(Pdb) found_bbb_ccc.bbb
<app.models.entities.bbb.Bbb object at 0x7f488dd73f10>
(Pdb) found_bbb_ccc.ccc
<app.models.entities.ccc.Ccc object at 0x7f488ce6ec40>
(Pdb) found_bbb_ccc.bbb.ccc
<app.models.entities.ccc.Ccc object at 0x7f488ce6ec40>
(Pdb) found_bbb_ccc.ccc.bbb
<app.models.entities.bbb.Bbb object at 0x7f488dd73f10>
```

This shows that a `bbb` can reference and find its `ccc` through the join model, and vice versa. Reading through the association is fine. But creating a new association by writing is problematic:

```python
new_bbb = Bbb(name='Bbb instance')
new_ccc = Ccc(name='Ccc instance')

new_bbb.ccc = new_ccc
db.session.commit()

*** sqlalchemy.exc.IntegrityError: (psycopg2.errors.NotNullViolation) null value in column "entity_type" violates not-null constraint
DETAIL:  Failing row contains (4b1f7ac7-16b0-4972-9577-bda1b5efe2aa, 2021-08-05 17:50:05.233465, 2021-08-05 17:50:05.233482, 63463492-0a9d-492f-b42a-72ec276f2768, null, a75d06af-33bd-4345-abbd-c6098e9a797d).

[SQL: INSERT INTO entity_ccc (created, updated, id, ccc_id, entity_id) VALUES (%(created)s, %(updated)s, %(id)s, %(ccc_id)s, %(entity_id)s)]
[parameters: {'created': datetime.datetime(2021, 8, 5, 17, 50, 5, 233465), 'updated': datetime.datetime(2021, 8, 5, 17, 50, 5, 233482), 'id': UUID('4b1f7ac7-16b0-4972-9577-bda1b5efe2aa'), 'ccc_id': UUID('a75d06af-33bd-4345-abbd-c6098e9a797d'), 'entity_id': UUID('63463492-0a9d-492f-b42a-72ec276f2768')}]
(Background on this error at: http://sqlalche.me/e/14/gkpj)
```

The error seen is that the writing of this data is not setting the polymorphic `entity_type`, which should be `bbb` here. I believe the problem is that the `secondary` relationship definition requires a table name, and not an object (passing `BbbCcc` would likely pick up the entity_type right?) but perhaps it is something else. 

How can I tweak this code to allow the setting of the polymorphic association as described? Thanks!

**Comment by zzzeek** [comment, marked as answer]
Hi I'm replying from my phone right now,.but the basic issue is your example is not making any use of this BbbCcc association entity.  Referring to the entity_ccc table has nothing to do with this ORM mapping, there is no "reverse" lookup from a plain Table to an arbitrary class that happens to be mapped to it.

The pattern you are looking to use is known as the association object pattern and you will see this at     https://docs.sqlalchemy.org/en/14/orm/basic_relationships.html#association-object .  You can of course shadow the association table by referring to it's table as "secondary" as you are doing but this does not pull in any ORM persistence features like polymorphic identity.   

Another option would be to set a default value for this column in entity_ccc but this would not have any polymorphic variability on persist.

**Reply by coneybeare** [reply]
Thanks for the reply! I gave this a shot as I was tinkering with the setup, but I ran into problems directly setting the entity directly like `ccc.bbb = bbb` and `bbb.ccc = ccc`. Perhaps I did something wrong, I'll try again and report back

**Reply by coneybeare** [reply]
I wasn't able to figure out how to combine that assocation object example with polymorphic columns. The closest I have found is the [discriminator_on_association.y](https://techspot.zzzeek.org/files/2007/discriminator_on_association.py) example from [here](https://techspot.zzzeek.org/2007/05/29/polymorphic-associations-with-sqlalchemy/#at-the-bottom), but that example is from from version 0.7 which is quite old

**Reply by zzzeek** [reply]
just create a new BbbCcc object, associate Bbb and Ccc with it, then persist.  

the discrimintator_on_assocaition example is maintained within SQLAlchemy's source distribution and documentation and you can see the current version here: https://docs.sqlalchemy.org/en/14/_modules/examples/generic_associations/discriminator_on_association.html

**Reply by coneybeare** [reply]
Ok, so putting all of this together and tinkering around quite a bit, it seems like I won't be able to get the syntax I desire with the polymorphic association table:
```
new_bbb = Bbb(name='Bbb instance')
new_ccc = Ccc(name='Ccc instance')
new_bbb.ccc = new_ccc
db.session.commit()
```

I was trying to keep the join model out of the syntax unless needed, but unless I am missing something, the polymorphism is problematic for the creation when going from one side of the association table to the other directly, ie: `bbb -> entity_ccc -> ccc` and vice_versa

**Reply by zzzeek** [reply]
once you have your assocaition object working you can then use the [association proxy](https://docs.sqlalchemy.org/en/14/orm/extensions/associationproxy.html) to provide the more succinct syntax where the association object is created automatically.   

if you can provide an [MCVE](http://stackoverflow.com/help/mcve) I can illustrate how this would work.

**Reply by coneybeare** [reply]
Thanks for the assist. Here is a streamlined project that has the above setup, with non-placeholder names.

```
pip install -r requirements.tx
bash -c "psql -U postgres postgres -c 'create database mcve;'"
alembic upgrade head
python -m pytest tests
```

When running the tests, two will fail due to the aforementioned issue, which I was not able to get to work with association proxy. Hopefully it is just some small oversight! Thank you for your help 👍 

[mcve.zip](https://github.com/sqlalchemy/sqlalchemy/files/6957330/mcve.zip)

**Reply by zzzeek** [reply]
hey there -

this is your whole app in a huge zip file, that's the opposite of "mcve".  please create a single file with just the bbb / ccc / bbbccc classes and show a simple example, just like the examples that come with SQLAlchemy etc.  thanks!

**Reply by coneybeare** [reply]
Sorry for the delayed response. I would still appreciate help if you are still willing! 
[mcve.py.zip](https://github.com/sqlalchemy/sqlalchemy/files/7008746/mcve.py.zip)

Unzip and `python mcve.py`. Print statements show expectations

**Reply by zzzeek** [reply]
ah OK the zip is just a .py file, strange, but easy enough

First fix the single-inheritance mapping you have:

```py
class BusinessUnitLegacyDatum(EntityLegacyDatum):

    # single table inheritance
    __tablename__ = None

    entity_id = sa.Column(
        sa.Integer,
        sa.ForeignKey(f"{BusinessUnit.__tablename__}.id"),
        nullable=False,
    )
    # ...

```

relationships that use entity_legacy_datum as "secondary" need to be "viewonly":

```py
    business_unit = sa.orm.relationship(
        "BusinessUnit",
        secondary="entity_legacy_datum",
        back_populates="legacy_datum",
        uselist=False,
        viewonly=True,
    )
```

action we want to perform is essentially:

```py
session.add(
   BusinessUnitLegacyDatum(
       business_unit=business_unit_2, legacy_datum=legacy_datum_2
   )
)
```

association proxy then looks like this:

```py
class BusinessUnit(Base):
   # ...

    legacy_datum_proxy = association_proxy(
        "business_unit_legacy_datum",
        "legacy_datum",
        creator=lambda legacy_datum: BusinessUnitLegacyDatum(
            legacy_datum=legacy_datum
        ),
    )
```

usage:

```py
business_unit_2.legacy_datum_proxy = legacy_datum_2

```

here's the full program:

```py
from datetime import datetime
from enum import Enum
import re

import sqlalchemy as sa
from sqlalchemy.ext.associationproxy import association_proxy
from sqlalchemy.ext.declarative import as_declarative
from sqlalchemy.ext.declarative import declared_attr
import sqlalchemy.orm as orm


class Timestamp(object):

    created = sa.Column(sa.DateTime, default=datetime.utcnow, nullable=False)
    updated = sa.Column(sa.DateTime, default=datetime.utcnow, nullable=False)


@as_declarative()
class Base(Timestamp):
    """SQLAlchemy Base class for all database models."""

    @declared_attr
    def __tablename__(cls) -> str:  # pylint: disable=no-self-argument
        """Define table name for all models as the snake case of the model's name."""
        first_pass = re.sub(
            r"(.)([A-Z][a-z]+)", r"\1_\2", cls.__name__
        )  # pylint: disable=no-member
        return re.sub(r"([a-z0-9])([A-Z])", r"\1_\2", first_pass).lower()

    id = sa.Column(sa.Integer, primary_key=True)


class Organization(Base):
    """Represents an Organization entity."""

    name = sa.Column(sa.TEXT)
    short_id = sa.Column(
        sa.Integer, sa.Sequence("organization_short_id_sequence")
    )


class BusinessUnit(Base):
    """Represents a Business Unit entity."""

    name = sa.Column(sa.TEXT)
    short_id = sa.Column(
        sa.Integer, sa.Sequence("business_unit_short_id_sequence")
    )

    business_unit_legacy_datum = sa.orm.relationship(
        "BusinessUnitLegacyDatum",
        back_populates="business_unit",
        uselist=False,
    )

    legacy_datum = sa.orm.relationship(
        "LegacyDatum",
        secondary="entity_legacy_datum",
        back_populates="business_unit",
        uselist=False,
    )

    legacy_datum_proxy = association_proxy(
        "business_unit_legacy_datum",
        "legacy_datum",
        creator=lambda legacy_datum: BusinessUnitLegacyDatum(
            legacy_datum=legacy_datum
        ),
    )


class LegacyDatum(Base):
    """Represents a Legacy Datum entity."""

    foo = sa.Column(sa.INTEGER, nullable=True)

    business_unit_legacy_datum = sa.orm.relationship(
        "BusinessUnitLegacyDatum", back_populates="legacy_datum", uselist=False
    )

    business_unit = sa.orm.relationship(
        "BusinessUnit",
        secondary="entity_legacy_datum",
        back_populates="legacy_datum",
        uselist=False,
        viewonly=True,
    )


class EntityTypes(Enum):
    """Enum values for all polymorphic Entity Types that TEvo supports."""

    ORGANIZATION = Organization.__name__.lower()
    BUSINESS_UNIT = BusinessUnit.__name__.lower()


class EntityLegacyDatum(Base):
    """1 to 1 polymorphic mapping between an EntityType and Legacy Datum."""

    entity_type = sa.Column(
        sa.Enum(EntityTypes, values_callable=lambda x: [e.value for e in x]),
        nullable=False,
    )

    legacy_datum_id = sa.Column(
        sa.Integer,
        sa.ForeignKey(f"{LegacyDatum.__tablename__}.id"),
        nullable=False,
    )

    __mapper_args__ = {"polymorphic_on": entity_type}


class BusinessUnitLegacyDatum(EntityLegacyDatum):
    """1 to 1 mapping between a BusinessUnit and Legacy Datum."""

    # single table inheritance
    __tablename__ = None

    entity_id = sa.Column(
        sa.Integer,
        sa.ForeignKey(f"{BusinessUnit.__tablename__}.id"),
        nullable=False,
    )

    business_unit = sa.orm.relationship(
        "BusinessUnit", back_populates="business_unit_legacy_datum"
    )
    legacy_datum = sa.orm.relationship(
        "LegacyDatum", back_populates="business_unit_legacy_datum"
    )

    __mapper_args__ = {
        "polymorphic_identity": EntityTypes(BusinessUnit.__name__.lower())
    }


engine = sa.create_engine("sqlite://", echo=True)
Base.metadata.create_all(engine)

session = orm.Session(engine)


# CREATE A RELATIONSHIP BY MODELING THE RELATIONSHIP FIRST AND ASSIGNING EXISTING MODELS
# TESTS READS ON THE ASSOCIATION TABLE
legacy_datum = LegacyDatum(foo=123)
business_unit = BusinessUnit(name="foobar")
business_unit_legacy_datum = BusinessUnitLegacyDatum(
    legacy_datum=legacy_datum, business_unit=business_unit
)
session.add_all([legacy_datum, business_unit, business_unit_legacy_datum])
session.commit()

session.query(BusinessUnit).filter(BusinessUnit.id == business_unit.id)

print("===============================================")
print(
    f"Business unit can read legacy datum with id: {business_unit.legacy_datum.id}"
)

session.query(LegacyDatum).filter(LegacyDatum.id == legacy_datum.id)

print("===============================================")
print(
    f"Legacy Datum can read business unit with id: {legacy_datum.business_unit.id}"
)


# CREATE A RELATIONSHIP BY ADDING AN ASSOCIATED MODEL DIRECTLY
# TESTS WRITES ON THE ASSOCIATION TABLE

legacy_datum_2 = LegacyDatum(foo=321)
business_unit_2 = BusinessUnit(name="barfoo")

session.add_all([legacy_datum_2, business_unit_2])
session.commit()

print("=================================================")
print("Attempting a direct set of the associated record")

# we want the equivalent of this
# session.add(
#    BusinessUnitLegacyDatum(
#        business_unit=business_unit_2, legacy_datum=legacy_datum_2
#    )
# )
business_unit_2.legacy_datum_proxy = legacy_datum_2
# legacy_datum_2.business_unit = business_unit_2
session.commit()
```

**Reply by coneybeare** [reply]
Thanks! Can the `creator` lambda make use of the String model syntax? This altered file works, but my project has classes split across files that cause a circular import here in the creator. Not sure if I can get past it without trickery

**Reply by CaselIT** [reply]
you can use any function in place of a lambda, so you can import what you need there
